pkg/ipam-node/cleaner/cleaner.go

```diff
@@ -0,0 +1,136 @@
+/*
+ Copyright 2023, NVIDIA CORPORATION & AFFILIATES
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+     http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package cleaner
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/google/uuid"
+	corev1 "k8s.io/api/core/v1"
+	apiErrors "k8s.io/apimachinery/pkg/api/errors"
+	apiTypes "k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	storePkg "github.com/Mellanox/nvidia-k8s-ipam/pkg/ipam-node/store"
+	"github.com/Mellanox/nvidia-k8s-ipam/pkg/ipam-node/types"
+)
+
+type Cleaner interface {
+	Start(ctx context.Context)
+}
+
+func New(client client.Client, storeMgr storePkg.Manager,
+	checkInterval time.Duration,
+	checkCountBeforeRelease int) Cleaner {
+	return &cleaner{
+		client:                  client,
+		storeMgr:                storeMgr,
+		checkInterval:           checkInterval,
+		checkCountBeforeRelease: checkCountBeforeRelease,
+		staleAllocations:        make(map[string]int),
+	}
+}
+
+type cleaner struct {
+	client                  client.Client
+	storeMgr                storePkg.Manager
+	checkInterval           time.Duration
+	checkCountBeforeRelease int
+	// key is <pool_name>|<container_id>|<interface_name>, value is count of failed checks
+	staleAllocations map[string]int
+}
+
+func (c *cleaner) Start(ctx context.Context) {
+	logger := logr.FromContextOrDiscard(ctx).WithName("cleaner")
+	for {
+		loopLogger := logger.WithValues("checkID", uuid.NewString())
+		loopLogger.Info("check for stale IPs")
+		ctx = logr.NewContext(ctx, loopLogger)
+		if err := c.loop(ctx); err != nil {
+			loopLogger.Error(err, "check failed")
+		}
+		select {
+		case <-ctx.Done():
+			logger.Info("shutdown cleaner")
+			return
+		case <-time.After(c.checkInterval):
+		}
+	}
+}
+
+func (c *cleaner) loop(ctx context.Context) error {
+	logger := logr.FromContextOrDiscard(ctx)
+	store, err := c.storeMgr.Open(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to open store: %v", err)
+	}
+	allReservations := map[string]struct{}{}
+	for _, poolName := range store.ListPools() {
+		for _, reservation := range store.ListReservations(poolName) {
+			resLogger := logger.WithValues("pool", poolName,
+				"container_id", reservation.ContainerID, "interface_name", reservation.InterfaceName)
+			key := c.getStaleAllocKey(poolName, reservation)
+			allReservations[key] = struct{}{}
+			if reservation.Metadata.PodName == "" || reservation.Metadata.PodNamespace == "" {
```

Comment:
pod uuid is not a required  metadata field for  cleaner loop ?

Reply:
Yes, it is not required. We use it as an extra check here. We can use POD_UID safely here only if it is received from a reliable source (as a part of CNI call). Suppose we add logic to query POD_UID from the k8s API (from the informer cache). In that case, we will need to completely remove POD_UID comparison here because POD_UID in the store may be invalid (it was read from an outdated cache), leading to the false cleanup of the IP.